--script and --lib flags

dsp opened the pull request:
- **fix missing space in init --build-system flag description**
  

- **init: add support for --bin**
   --lib and --lib-rs,This adds support for three new flags which determine the templates
  we generate:
  
 1. We adding --lib, which will generate the standard python library
       We no longer generate `[project.scripts]` section and updated
       the template.
2. We added --script, which will generate a package with a __main__.py,
       __init__.py and an entry to `[project.scripts]`. `__main__.py` follows
       semantics found in Python packages such as `ensurepip` and `pip.
3. We added simple tests. Tests do not cover yet `--lib` and `--build-system=maturin`.
4. We added a short section about executable projects to docs/guide/basic.md
5. We moved all in init.rs to a `templates` folder and use `include_str!()`.

dsp commented:
Adding this as a draft to get feedback, while I am working to add tests.

dsp commented:
> I think this looks good, I'm however not sure if `--lib-rs` is a good idea. I do however also recognize that this build system thing is odd / verbose.

My reasoning here is that these options determine the template used. The intuition here is that I likely know the kind of project i want, but leave the specifics of build system choice, etc to rye to choose. This means that rye has "sane default' switch for a template. For instance --lib and --bin implying hatchling and --lib-rs implying maturin. If in the future some new build system comes out and rye chooses this, we can flip the switch to have --lib-rs imply something else.

dsp commented:
The counter argument btw is that we want to implment a PyO3/Maturin template for `--bin`  as well. Instead of then having an awkward `--bin-rs` we just have any combination of `--build-system` and `--bin` and `--lib` flags.

So our choices are effectively:
1. Only keep --bin and --lib, build system choises the specific template choosen.
2. Add --lib-rs and treat them all as explicit template choices. This means later we might need --bin-rs, etc.
3. Some alternative solution (e.g. --lib=rust)

mitsuhiko commented:
I would go with `--lib` and `--bin` and then use the build system for the template. It can always error out and say it's not a sensible combination. I also agree that `--bin` is weird. Maybe it should be called `--script` instead?

dsp commented:
- added tests.
- added changelog entry.
- added an entry to docs/guide/basic.md
- refactored how we deal with templates in `init.rs` (this is deliberately the last commit and can be dropped, if needed)

mitsuhiko commented on rye/src/templates/script/default/__main__.py:
```diff
@@ -0,0 +1,4 @@
+import {{ name_safe }}
```
I know people love `__main__` but do we think it's a good idea? At least personally I really try to avoid it. Not sure how others feel about it.

dsp replied:
I looked into common Python packages like ensurepip, pip, etc. They all provide `__main__.py` so that you can do `python -m pip` for example. I think it is good to provide a way so people can run `python -m my-package` from a rye project. That's why i included it as a thin wrapper around __init__

mitsuhiko commented on rye/src/cli/init.rs:
```diff
@@ -82,193 +85,54 @@ pub struct Args {
-    def __getattr__(name):
-        return getattr(__setuptools, name)
-"#;
+const SETUP_PY_PROXY_SCRIPT: &str = include_str!("../templates/setuptools.py");
```
I don't think this should be in `templates` since it's not a template.

dsp replied:
Fair enough will do.

mitsuhiko commented on rye/src/templates/LICENSE.txt:
```diff
@@ -0,0 +1 @@
+{{ license_text }}
```
I would add `.j2` to all templates as extension for syntax highlighting in editors.

dsp replied:
Sounds reasonable. I looked into the examples directory of minijinja as a reference and they didn't have `.j2`, but it makes sense to add it.

mitsuhiko commented:
This looks good!
